src/cairo_run.rs

```diff
@@ -0,0 +1,141 @@
+use crate::{utils::to_py_error, vm_core::PyVM};
+use cairo_rs::{
+    cairo_run::write_output,
+    hint_processor::builtin_hint_processor::builtin_hint_processor_definition::BuiltinHintProcessor,
+    types::{program::Program, relocatable::Relocatable},
+    vm::{
+        errors::{
+            cairo_run_errors::CairoRunError, runner_errors::RunnerError, trace_errors::TraceError,
+            vm_errors::VirtualMachineError,
+        },
+        runners::cairo_runner::CairoRunner,
+    },
+};
+use pyo3::{pyfunction, PyResult};
+use std::path::{Path, PathBuf};
+
+#[pyfunction]
+#[pyo3(name = "cairo_run")]
+pub fn cairo_run_py<'a>(
+    path: &'a str,
+    entrypoint: &'a str,
+    trace_enabled: bool,
+    print_output: bool,
+    trace_file: Option<&str>,
+    memory_file: Option<&str>,
+) -> PyResult<()> {
+    let path = Path::new(path);
+    let program = Program::new(path, entrypoint).map_err(to_py_error)?;
+    let hint_processor = BuiltinHintProcessor::new_empty();
+    let mut cairo_runner = CairoRunner::new(&program, &hint_processor).map_err(to_py_error)?;
+    let vm = PyVM::new(program.prime, trace_enabled);
+    let end = cairo_runner
+        .initialize(&mut vm.vm.borrow_mut())
+        .map_err(to_py_error)?;
+
+    run_until_pc(&mut cairo_runner, end, &vm).map_err(to_py_error)?;
+
+    vm.vm
+        .borrow_mut()
+        .verify_auto_deductions()
+        .map_err(to_py_error)?;
+
+    cairo_runner
+        .relocate(&mut vm.vm.borrow_mut())
+        .map_err(to_py_error)?;
+
+    if print_output {
+        write_output(&mut cairo_runner, &mut vm.vm.borrow_mut()).map_err(to_py_error)?;
+    }
+
+    if let Some(trace_path) = trace_file {
+        let trace_path = PathBuf::from(trace_path);
+        let relocated_trace = cairo_runner
+            .relocated_trace
+            .as_ref()
+            .ok_or(CairoRunError::Trace(TraceError::TraceNotEnabled))
+            .map_err(to_py_error)?;
+
+        match cairo_rs::cairo_run::write_binary_trace(relocated_trace, &trace_path) {
+            Ok(()) => (),
+            Err(_e) => {
+                return Err(CairoRunError::Runner(RunnerError::WriteFail)).map_err(to_py_error)
+            }
+        }
+    }
+
+    if let Some(memory_path) = memory_file {
+        let memory_path = PathBuf::from(memory_path);
+        match cairo_rs::cairo_run::write_binary_memory(&cairo_runner.relocated_memory, &memory_path)
+        {
+            Ok(()) => (),
+            Err(_e) => {
+                return Err(CairoRunError::Runner(RunnerError::WriteFail)).map_err(to_py_error)
+            }
+        }
+    }
+
+    Ok(())
```

Comment:
Could something like this be done?
i```
f let Some(memory_path) = memory_file {
    let memory_path = PathBuf::from(memory_path);
    cairo_rs::cairo_run::write_binary_memory(&cairo_runner.relocated_memory, &memory_path)
        .map_err(|_| to_py_error(CairoRunError::Runner(RunnerError::WriteFail)))
else {
    Ok(())
}
```

Reply:
nice catch! solved in 19b5dfe47f65cbf7b535e6e31078bb3b4cb06693